ASoC: SOF: debug: show firmware/topology prefix/names

plbossart opened the pull request:
We currently log the firmware and topology names in the kernel logs, but there's been an ask to add the information in debugfs to simplify CI scripts.

This isn't meant to be a stable ABI used by apps, changes will be allowed as needed.

Closes: https://github.com/thesofproject/linux/issues/3867

plbossart committed:
ASoC: SOF: debug: show firmware/topology prefix/names

The SOF driver has multiple profiles to select firmware/topology
prefix/names depending on the platform and ipc_type, and each of those
fields can be overridden with kernel parameters. This results in some
cases in confusion on what configuration is actually used in a given
test.

We currently log the firmware and topology names in the kernel logs,
but there's been an ask to add the information in debugfs to simplify
test scripts used by developers and CI.

This isn't meant to be a stable ABI used by apps, changes will be
allowed as needed.

Closes: https://github.com/thesofproject/linux/issues/3867
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

marc-hb reviewed (dismissed):
Lovely

ujfalusi commented on sound/soc/sof/debug.c:
```diff
@@ -338,6 +339,13 @@ int snd_sof_dbg_init(struct snd_sof_dev *sdev)
+	debugfs_create_str("tplg_path", 0444, sdev->debugfs_root, (char **)&plat_data->tplg_filename_prefix);
+	debugfs_create_str("fw_name", 0444, sdev->debugfs_root, (char **)&plat_data->fw_filename);
+	debugfs_create_str("tplg_name", 0444, sdev->debugfs_root, (char **)&plat_data->tplg_filename);
+	debugfs_create_u32("ipc_type", 0444, sdev->debugfs_root, (u32 *)&plat_data->ipc_type);
```
Should we have a folder to hold these? /sys/kernel/debug/sof/ starts to be a bit busy.

plbossart replied:
@ujfalusi we can add a folder, but what would the name be? 'fw_profile' or something?

ujfalusi replied:
@plbossart, you got me ;)
it is not just paths and filenames but the ipc_type as well and it is not really a FW configuration.
`profile` or `driver profile`, these dictates what firmware and what IPC to use. `platform profile`?

I'm really bad at this.

plbossart replied:
"ci_profile" ?

ujfalusi replied:
it is not a ci profile, I would say, it is a platform_profile, or platform_config?
On cAVS2.5 we can have IPC3 or IPC4 depending on what is selected, selected_config, selected_profile?

marc-hb replied:
Both `fw_config` and `fw_profile` work for me.

`fw_path` is technically not part of the firmware configuration, but the reason why we have different `fw_path` is because they point to different firmware configurations. For instance, some of our firmware files have only a signing key difference.

plbossart replied:
Any objections on "fw_profile" then?

marc-hb commented on sound/soc/sof/debug.c:
```diff
@@ -330,14 +330,26 @@ EXPORT_SYMBOL_GPL(snd_sof_dbg_memory_info_init);
 	/* use "sof" as top level debugFS dir */
 	sdev->debugfs_root = debugfs_create_dir("sof", NULL);
 
+	/* expose firmware/topology prefix/names for CI integration */
```
```suggestion
	/* expose firmware/topology prefix/names for test purposes */
```

marc-hb reviewed (dismissed):
>  an ask to add the information in debugfs to simplify CI scripts.

The TEST scripts, NOT the CI scripts.

= the scripts that EVERYONE should run, not the obscure Jenkins scripts or Github workflows no one knows about.

I mean this PR is not to please one particular and demanding user. It's useful for everyone interested in testing their code (= too few people?)

plbossart committed:
ASoC: SOF: Intel: add default firmware library path for LNL

The commit cd6f2a2e6346 ("ASoC: SOF: Intel: Set the default firmware
library path for IPC4") added the default_lib_path field for all
platforms, but this was missed when LunarLake was later introduced.

Fixes: 64a63d9914a5 ("ASoC: SOF: Intel: LNL: Add support for Lunarlake platform")
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
@ujfalusi This PR was at least useful to detect that we never had a default_lib_path for LNL, see second commit...

ujfalusi commented:
@plbossart, one checkpatch issue should be addressed.

plbossart commented:
> @plbossart, one checkpatch issue should be addressed.

Done. I didn't realize my post-commit hook was missing.

marc-hb commented on sound/soc/sof/debug.c:
```diff
@@ -330,14 +330,32 @@ EXPORT_SYMBOL_GPL(snd_sof_dbg_memory_info_init);
+	debugfs_create_str("fw_path", 0444, fw_profile,
+			   (char **)&plat_data->fw_filename_prefix);
+	debugfs_create_str("fw_lib_path", 0444, fw_profile,
+			   (char **)&plat_data->fw_lib_prefix);
```
There is page_fault_oops with `fw_lib_path` in some configurations
Found by... testing!
- https://github.com/thesofproject/sof-test/pull/1178


```
root@jf-cml-rvp-sdw-3:~# cat /sys/kernel/debug/sof/fw_profile/fw_lib_path
Killed

Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: BUG: kernel NULL pointer dereference, address: 0000000000000000
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: #PF: supervisor read access in kernel mode
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: #PF: error_code(0x0000) - not-present page
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: PGD 0 P4D 0
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: Oops: 0000 [#5] PREEMPT SMP NOPTI
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: CPU: 8 PID: 2096 Comm: grep Tainted: G      D            6.9.0-rc3-g0c12720aaa21 #dev
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: Hardware name: Intel Corporation CometLake Client Platform/CometLake U DDR4 HR RVP, BIOS CMLSFWR1.R00.1245.D00.1906>
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: RIP: 0010:strlen+0x4/0x30
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: Code: f7 75 ec 31 c0 c3 cc cc cc cc 48 89 f8 c3 cc cc cc cc 0f 1f 40 00 90 90 90 90 90 90 90 90 90 90 90 90 90 90 9>
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: RSP: 0018:ffffaa4b438c7e28 EFLAGS: 00010246
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: RAX: ffff8d2750c389b8 RBX: ffff8d2743c6b800 RCX: 0000000000000006
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: RDX: 0000000000000007 RSI: 0000558abe473000 RDI: 0000000000000000
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: RBP: ffff8d2743c6b800 R08: 0000000000000000 R09: 0000000000000000
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: R10: 0000000000000000 R11: 0000000000000000 R12: 0000000000000000
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: R13: 0000000000018000 R14: ffffaa4b438c7f08 R15: ffff8d275f0ce8f8
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: FS:  00007f942669e740(0000) GS:ffff8d2893600000(0000) knlGS:0000000000000000
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: CR2: 0000000000000000 CR3: 000000013bdf6004 CR4: 00000000003706f0
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel: Call Trace:
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  <TASK>
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ? __die+0x24/0x70
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ? page_fault_oops+0x15a/0x450
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ? exc_page_fault+0x68/0x1e0
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ? asm_exc_page_fault+0x26/0x30
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ? strlen+0x4/0x30
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  debugfs_read_file_str+0x4d/0xf0
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  vfs_read+0xab/0x340
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ? __do_sys_newfstatat+0x35/0x60
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  ksys_read+0x69/0xf0
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  do_syscall_64+0xad/0x1b0
Apr 22 23:33:23 jf-cml-rvp-sdw-3 kernel:  entry_SYSCALL_64_after_hwframe+0x72/0x7a

```

plbossart replied:
this is a CML platform....

I can't recall is fw_lib_path is relevant for IPC3 @ujfalusi, we could just skip the file create for IPC3 I guess.

At any rate the scripts only used this for IPC4 platforms.

marc-hb replied:
Yes I confirm the oops happened only with the stable-v2.2 firmware branch. But it did happen with ALL 3 devices tested for that branch.

Because of various sudo issues (hopefully last fix in https://github.com/thesofproject/sof-test/pull/1178) `fw_lib_path` had never been tested yet.

ujfalusi replied:
Yes, fw_lib_path is going to be NULL with IPC3, probably we can have have a static const "N/A" string here and give that to `debugfs_create_str()`?
Or skip creating the file in case of IPC3, it is not applicable after all.

plbossart replied:
@marc-hb what's the preferred result for scripts? A N/A reply or the file not created at all?

marc-hb replied:
Either can be scripted easily but no file at all is _safer_ because there's less risk of confusing it with something valid:

https://en.wikipedia.org/wiki/Null_pointer  a.k.a. the "billion dollar mistake".

Note we have no script trying to read this brand new field yet (= how I just found this bug)
